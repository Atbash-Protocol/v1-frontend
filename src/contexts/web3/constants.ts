export const USER_REJECTED = "User Rejected";
