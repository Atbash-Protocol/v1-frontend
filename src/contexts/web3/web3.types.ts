import { providers } from 'ethers';
import Web3Modal from 'web3modal';

export interface WEB3State {
    networkID: number | null;
    provider: providers.JsonRpcProvider | null;
    signer: providers.Web3Provider | null;
    signerAddress: string | null;
    web3Modal: Web3Modal;
}

export interface IWeb3Context {
    state: WEB3State;
    memoConnect: () => void;
    memoDisconnect: (currentSigner: WEB3State['signer']) => Promise<void>;
}

export enum WEB3ActionTypesEnum {
    CLOSE = 'CLOSE',
    NETWORK_CHANGED = 'NETWORK_CHANGED',
    SET_SIGNER = 'SET_SIGNER',
    CLEAR_SIGNER = 'CLEAR_SIGNER',
    SET_PROVIDER = 'SET_PROVIDER',
}

export interface WEB3ContextAction {
    type: WEB3ActionTypesEnum;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    payload?: any; // TODO : better typing
}
