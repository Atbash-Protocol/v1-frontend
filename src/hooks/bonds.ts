import { useMemo } from 'react';

import { useSelector } from 'react-redux';

import { BondItem } from 'store/modules/bonds/bonds.types';
import { IReduxState } from 'store/slices/state.interface';

// Slaps together bond data within the account & bonding states
const useBonds = () => {
    const bonds = useSelector<IReduxState, BondItem[]>(state => Object.values(state.bonds.bonds));

    const mostProfitableBonds = useMemo(
        () =>
            bonds.sort((bond1, bond2): number => {
                if (bond1.metrics.bondDiscount === null || bond2.metrics.bondDiscount === null) return 0;

                return bond1.metrics.bondDiscount > bond2.metrics.bondDiscount ? 0 : 1;
            }),
        [bonds],
    );

    return {
        bonds,
        mostProfitableBonds,
    };
};

export default useBonds;

export const selectBondReady = (bond: BondItem) => {
    return Object.values(bond.metrics).some(metric => metric !== null);
};

export const useBondPurchaseReady = () => {
    const metrics = useSelector<IReduxState, boolean>(state => state.main.metrics.reserves !== null);

    const marketReady = useSelector<IReduxState, boolean>(state => state.markets.markets.dai !== null);

    return metrics && marketReady;
};
