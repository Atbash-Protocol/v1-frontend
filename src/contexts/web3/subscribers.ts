import { Dispatch } from 'react';

import { providers } from 'ethers';
import Core from 'web3modal';

import { DEFAULT_NETWORK } from 'constants/blockchain';
import { getProviderURI } from 'contexts/web3/web3.utils';

import { WEB3ContextAction, WEB3ActionTypesEnum } from './web3.types';

export const subscribeSigner = async (web3provider: providers.Web3Provider, dispatch: Dispatch<WEB3ContextAction>) => {
    web3provider.on('networkChanged', async (networkId: number) => {
        console.log('sub', networkId);
        const signer = new providers.Web3Provider(web3provider as unknown as providers.ExternalProvider);

        dispatch({ type: WEB3ActionTypesEnum.NETWORK_CHANGED, payload: { signer, networkId } });
    });

    web3provider.once('close', () => {
        dispatch({ type: WEB3ActionTypesEnum.CLOSE, payload: { signer: web3provider.getSigner() } });
    });
};

export const createSigner = async (web3Modal: Core, dispatch: Dispatch<WEB3ContextAction>) => {
    const web3Provider = await web3Modal.connect();

    const signer = new providers.Web3Provider(web3Provider);

    subscribeSigner(web3Provider, dispatch);

    try {
        const [{ chainId }, address] = await Promise.all([signer.getNetwork(), signer.getSigner().getAddress()]);

        dispatch({ type: WEB3ActionTypesEnum.SET_SIGNER, payload: { signer, chainId, address } });
    } catch (err) {
        console.error(err);
    } finally {
    }
};

export const subscribeProvider = async (dispatch: Dispatch<WEB3ContextAction>) => {
    const provider = new providers.JsonRpcProvider(getProviderURI(DEFAULT_NETWORK));
    provider.on('error', err => {
        alert('provider error');
        console.error(err);
    });

    const { chainId } = await provider.ready;

    if (!chainId) throw new Error('Impossible to connect to the given provider');

    dispatch({ type: WEB3ActionTypesEnum.SET_PROVIDER, payload: { chainId, provider } });
};

export const resetWeb3Signer = (dispatch: Dispatch<WEB3ContextAction>, web3Modal: Core): void => {
    web3Modal.clearCachedProvider();
    localStorage.setItem('WEB3_CONNECT_CACHED_PROVIDER', '');
    dispatch({ type: WEB3ActionTypesEnum.CLOSE, payload: null });
};
