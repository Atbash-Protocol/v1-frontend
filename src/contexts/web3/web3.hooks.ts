import { useContext, useEffect, useMemo, useState } from 'react';

import { Dispatch } from 'redux';

import { DEFAULT_NETWORK } from 'constants/blockchain';
import { Web3Context } from 'contexts/web3/web3.context';
import { walletConnectWarning } from 'store/modules/messages/messages.slice';

export const useGoodNetworkCheck = () => {
    const {
        state: { networkID },
    } = useContext(Web3Context);

    const [isNetworkOk, setIsNetworkOk] = useState(false);

    useEffect(() => {
        if (networkID) setIsNetworkOk(networkID === DEFAULT_NETWORK);
    }, [networkID]);

    return isNetworkOk;
};

export const useSignerConnected = () => {
    const {
        state: { signer, networkID },
    } = useContext(Web3Context);

    const [isConnected, setIsConnected] = useState(signer !== null);

    useEffect(() => {
        if (signer && networkID) {
            setIsConnected(true);
        }
    }, [isConnected, signer, networkID]);

    return isConnected;
};

export const useWeb3ContextInitialized = () => {
    const {
        state: { signer, networkID, provider },
    } = useContext(Web3Context);

    const [isContextInitialized, setIsContextInitialized] = useState([provider, signer, networkID].some(e => e !== null));

    useEffect(() => {
        if (provider || signer) {
            setIsContextInitialized(true);
        }
    }, [provider, signer]);

    return isContextInitialized;
};

export const useSignerAddress = () => {
    const {
        state: { signerAddress },
    } = useContext(Web3Context);

    const [signer, setSigner] = useState(signerAddress);

    useEffect(() => {
        if (signerAddress) {
            setSigner(signerAddress);
        }
    }, [signerAddress]);

    return signer;
};

export const useProvider = () => {
    const {
        state: { provider },
    } = useContext(Web3Context);

    return provider;
};

export const useSafeSigner = (dispatch?: Dispatch) => {
    const {
        state: { signer, signerAddress },
    } = useContext(Web3Context);

    const memoSigner = useMemo(() => {
        if (!signer || !signerAddress) {
            if (dispatch) dispatch(walletConnectWarning);

            throw new Error('Missing signer or signerAddress ');
        }

        return { signer, signerAddress };
    }, [signer, signerAddress]);

    return memoSigner;
};
