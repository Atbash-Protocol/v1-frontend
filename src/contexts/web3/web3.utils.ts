import WalletConnectProvider from '@walletconnect/web3-provider';
import Web3Modal from 'web3modal';

import { Networks } from 'constants/blockchain';
import { theme } from 'constants/theme';
import { getProviderURL } from 'lib/contracts/networks';

export const initWeb3Modal = () =>
    new Web3Modal({
        network: '0x1',
        cacheProvider: true,
        theme: {
            background: theme.palette.cardBackground.main,
            main: theme.palette.primary.light,
            secondary: theme.palette.primary.light,
            hover: theme.palette.grey[800],
            border: 'none',
        },

        providerOptions: {
            walletconnect: {
                package: WalletConnectProvider,
                options: {
                    rpc: {
                        [Networks.LOCAL]: getProviderURL(Networks.LOCAL),
                        [Networks.RINKEBY]: getProviderURL(Networks.RINKEBY),
                        [Networks.MAINNET]: getProviderURL(Networks.MAINNET),
                    },
                },
            },
        },
    });

export const getProviderURI = (networkId: number): string => {
    if (![Networks.MAINNET, Networks.RINKEBY, Networks.LOCAL].includes(networkId)) throw new Error(`Unable to get mainNetURI with network "${networkId}"`);

    const rpcUrl: string | undefined = process.env.REACT_APP_NETWORK_RPC_URL;
    if (!rpcUrl) throw new Error('Network RPC URL not defined');

    return rpcUrl;
};
