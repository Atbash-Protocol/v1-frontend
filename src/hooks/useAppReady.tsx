import { useContext, useState, useLayoutEffect } from 'react';

import { Web3Context } from 'contexts/web3/web3.context';

export const useAppReady = () => {
    const {
        state: { signer, networkID, provider },
    } = useContext(Web3Context);

    const [isContextInitialized, setIsContextInitialized] = useState([provider, signer, networkID].some(e => e !== null));

    useLayoutEffect(() => {
        if (provider || signer) {
            setIsContextInitialized(true);
        }
    }, [provider, signer]);

    return isContextInitialized;
};
