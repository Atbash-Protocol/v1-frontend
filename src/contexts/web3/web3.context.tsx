import { createContext, useCallback, useContext, useEffect, useReducer } from 'react';

import { initWeb3Modal } from 'contexts/web3/web3.utils';

import { USER_REJECTED } from './constants';
import { subscribeProvider, createSigner, resetWeb3Signer } from './subscribers';
import { WEB3ActionTypesEnum, Web3Context, WEB3ContextAction, WEB3State } from './web3.types';

export const initialState: WEB3State = {
    provider: null,
    signer: null,
    signerAddress: null,
    networkID: null,
    web3Modal: initWeb3Modal(),
};

export const PWeb3Context = createContext<Web3Context>({
    state: initialState,
    memoConnect: Promise.reject,
    memoDisconnect: Promise.reject,
});

export const WEB3Reducer = (state: WEB3State, action: WEB3ContextAction) => {
    switch (action.type) {
        case WEB3ActionTypesEnum.CLOSE:
            return { ...state, signer: null, web3Modal: initWeb3Modal() };
        case WEB3ActionTypesEnum.NETWORK_CHANGED:
            return { ...state, networkID: Number(action.payload) };
        case WEB3ActionTypesEnum.SET_SIGNER:
            return { ...state, signer: action.payload.signer, signerAddress: action.payload.address, networkID: action.payload.chainId };
        case WEB3ActionTypesEnum.SET_PROVIDER:
            return { ...state, networkID: action.payload.chainId, provider: action.payload.provider };
        default:
            return state;
    }
};

export const NewWeb3ContextProvider = ({ children }: { children: JSX.Element }) => {
    const {
        state: { web3Modal, signer, provider },
    } = useContext(PWeb3Context);
    const [state, dispatch] = useReducer(WEB3Reducer, initialState);

    const memoConnect = useCallback(async () => {
        if (web3Modal && !signer) {
            await createSigner(web3Modal, dispatch);
        }
    }, [signer, web3Modal]);

    const memoDisconnect = useCallback(
        async (currentSigner: WEB3State['signer']) => {
            if (currentSigner && web3Modal) await resetWeb3Signer(dispatch, web3Modal);
        },
        [web3Modal],
    );

    useEffect(() => {
        // happens when user reload when metamask popup appears
        if (web3Modal) {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            web3Modal.on('error', (err: any) => {
                if (err && err?.message && err?.message === USER_REJECTED) alert('User rejected connection');
            });
        }

        // forces the connection
        if (web3Modal && web3Modal.cachedProvider?.length > 0) {
            memoConnect();
        } else {
            if (!provider) {
                subscribeProvider(dispatch); // init at the provider
            }
        }
    }, [web3Modal]);

    // useEffect(() => {
    //     if (!signer && !provider) {
    //         console.log("Reaching this context");
    //         subscribeProvider(dispatch); // init at the provider
    //     }
    // }, [provider, signer]);

    return (
        <PWeb3Context.Provider
            value={{
                state,
                memoConnect,
                memoDisconnect,
            }}
        >
            {children}
        </PWeb3Context.Provider>
    );
};

export const usePWeb3Context = () => {
    return useContext(PWeb3Context);
};
