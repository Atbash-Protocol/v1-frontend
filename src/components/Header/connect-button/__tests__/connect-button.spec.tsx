import { configureStore } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';

import { Web3Context } from 'contexts/web3/web3.context';
import * as WEB3HookModules from 'contexts/web3/web3.hooks';
import pendingTransactionsReducer from 'store/slices/pending-txns-slice';

import ConnectMenu from '../index';

function renderComponent(component: JSX.Element, contextState?: any) {
    return render(
        <Web3Context.Provider value={contextState}>
            <Provider
                store={configureStore({
                    reducer: {
                        pendingTransactions: pendingTransactionsReducer,
                    },
                })}
            >
                {component}
            </Provider>
            ,
        </Web3Context.Provider>,
    );
}

describe('ConnectMenu', () => {
    describe('when user is disconnected', () => {
        it('triggers a connect on click if disconnected', () => {
            const memoConnect = jest.fn();
            const { container } = renderComponent(<ConnectMenu />, {
                state: {
                    provider: null,
                    signer: null,
                },
                memoConnect,
                memoDisconnect: jest.fn(),
            });

            const connectButton = container.querySelector('button');
            expect(container.querySelector('button')?.textContent).toEqual('ConnectWallet');

            connectButton!.dispatchEvent(new MouseEvent('click', { bubbles: true }));
            expect(memoConnect).toHaveBeenCalledTimes(1);

            expect(container).toMatchSnapshot()
        });
    });

    describe('when user is connected', () => {
        it('shows indicates a wrong network', () => {
            jest.spyOn(WEB3HookModules, 'useGoodNetworkCheck').mockReturnValue(false);
            const { container } = renderComponent(<ConnectMenu />, {
                state: {
                    provider: null,
                    signer: 'signer',
                    signerAddress: 'signerAddress',
                },
            });

            expect(container.querySelector('button')?.textContent).toEqual('WrongNetwork');
        });

        it('triggers a connect on click if disconnected', () => {
            const memoDisconnect = jest.fn();
            jest.spyOn(WEB3HookModules, 'useGoodNetworkCheck').mockReturnValue(true);

            const { container } = renderComponent(<ConnectMenu />, {
                state: {
                    provider: null,
                    signer: 'signer',
                    signerAddress: 'signerAddress',
                },
                memoDisconnect,
            });

            const connectButton = container.querySelector('button');
            expect(container.querySelector('button')?.textContent).toEqual('Disconnect');

            connectButton!.dispatchEvent(new MouseEvent('click', { bubbles: true }));
            expect(memoDisconnect).toHaveBeenCalledTimes(1);
        });
    });
});
