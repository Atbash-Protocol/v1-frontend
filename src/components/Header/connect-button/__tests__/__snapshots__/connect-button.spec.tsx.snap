// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectMenu when user is disconnected triggers a connect on click if disconnected 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <button
      class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-575ak1-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Connect Wallet
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  ,
</div>
`;
